### PR TITLE
Spec: Add enableDebugMode()

### PR DESCRIPTION
**Pull request by alexmturner**
Introduces the debug mode to the spec and the concept of a 'debug scope'

***
<a href="https://pr-preview.s3.amazonaws.com/patcg-individual-drafts/private-aggregation-api/pull/60.html" title="Last updated on Jun 13, 2023, 6:14 PM UTC (8d29739)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/patcg-individual-drafts/private-aggregation-api/60/e79ff12...8d29739.html" title="Last updated on Jun 13, 2023, 6:14 PM UTC (8d29739)">Diff</a>

### CODE REVIEW

**Comment by alexmturner** [comment]
@linnan-github, could you PTAL? Thanks!
cc @yoavweiss

File: spec.bs [hunk 1]
```diff
@@ -194,6 +230,16 @@ Note: This last requirement means that global scopes with different origins
 
     Issue: Link Same-origin policy section.
 
+For any [=debug scope=] returned by the [=scoping details/get debug scope
+steps=], the [=mark a debug scope complete=] steps should later be performed
+given that same debug scope.
```
Comment (linnan-github) [inline spec.bs:235]:
```suggestion
given that same [=debug scope=].
```

File: spec.bs [hunk 2]
```diff
@@ -208,13 +254,39 @@ report=].
...
+
 <h4 dfn-type=dfn>Contribution cache entry</h4>
 A contribution cache entry is a [=struct=] with the following items:
 <dl dfn-for="contribution cache entry">
 : <dfn>contribution</dfn>
 :: A {{PAHistogramContribution}}
 : <dfn>batching scope</dfn>
 :: A [=batching scope=]
+: <dfn>debug info</dfn>
+:: A [=debug scope=] or a [=debug details=]
```
Comment (linnan-github) [inline spec.bs:289]: Wdyt if we have separate fields for debug scope and debug details to be more clear?

File: spec.bs [hunk 3]
```diff
@@ -297,28 +374,63 @@ To <dfn algorithm export>append an entry to the contribution cache</dfn> given a
 [=contribution cache entry=] |entry|:
 1. [=list/Append=] |entry| to the [=contribution cache=].
 
+To <dfn algorithm export>mark a debug scope complete</dfn> given a [=debug
+scope=] |debugScope|:
+
+1. Let |debugDetails| be a new [=debug details=] with the following items:
+    : [=debug details/enabled=]
```
Comment (linnan-github) [inline spec.bs:381]: Do we have to call out these default values?

File: spec.bs [hunk 4]
```diff
@@ -630,14 +755,17 @@ after "Let |operation| be |operationMap|[|name|]." (renumbering later steps as
 appropriate):
 <div algorithm="shared-storage-run-monkey-patch-1">
 2. Let <var ignore>batchingScope</var> be a new [=batching scope=].
+1. Let <var ignore>debugScope</var> be a new [=debug scope=].
```
Comment (linnan-github) [inline spec.bs:758]: Should this be step 3?

File: spec.bs [hunk 5]
```diff
@@ -670,15 +800,20 @@ as appropriate:
...
+        : [=scoping details/get batching scope steps=]
+        :: An algorithm that returns the [=batching scope=] that is scheduled to
+            be passed to [=process contributions for a batching scope=] when the
+            call currently executing in |scope| returns.
+        : [=scoping details/get debug scope steps=]
+        :: An algorithm that returns the [=debug scope=] that is scheduled to be
+            passed to [=mark a debug scope complete=] when the call currently
+            executing in |scope| returns.
+
+        Note: Multiple operation invocations can be in-progress at the same
```
Comment (linnan-github) [inline spec.bs:814]: Is this argument the same for debug scope?

Comment (alexmturner) [reply spec.bs:814]: Yep, edited to reflect that

File: spec.bs [hunk 6]
```diff
@@ -399,8 +397,9 @@ a [=batching scope=] |batchingScope|, an [=origin=] |reportingOrigin| and a
 1. [=list/iterate|For each=] |entry| of the [=contribution cache=]:
     1. If |entry|'s [=contribution cache entry/batching scope=] is
         |batchingScope|:
-        1. [=Assert=] |entry|'s [=contribution cache entry/debug info=] is a
-            [=debug details=].
+        1. [=Assert=] |entry|'s [=contribution cache entry/debug details=] is
```
Comment (linnan-github) [inline spec.bs:400]:
```suggestion
        1. [=Assert=]: |entry|'s [=contribution cache entry/debug details=] is
```

File: spec.bs [hunk 7]
```diff
@@ -1092,10 +1092,9 @@ stored for the [=consume budget if permitted=] algorithm contain data about a
 user’s web activity. As such, user controls to delete this data are required,
 see [clearing storage](#clearing-storage).
 
-On the other hand, the [=contribution cache=] only contains short-lived data
-tied to particular batching scopes, so controls are not required.
-
-Issue: Link batching scopes
+On the other hand, the [=contribution cache=] and [=debug scope map=] only
+contain short-lived data tied to particular [=batching scopes=], so controls are
```
Comment (linnan-github) [inline spec.bs:1096]: Maybe tied to particular batching scopes or debug scopes?

File: spec.bs [hunk 8]
```diff
@@ -297,28 +374,63 @@ To <dfn algorithm export>append an entry to the contribution cache</dfn> given a
...
+    1. If |debugDetails|'s [=debug details/key=] is not null, [=assert=]
+        |debugDetails|'s [=debug details/enabled=] is true.
+1. [=list/iterate|For each=] |entry| of the [=contribution cache=]:
+    1. If |entry|'s [=contribution cache entry/debug info=] is |debugScope|,
+        set |entry|'s [=contribution cache entry/debug info=] to |debugDetails|.
+
 To <dfn algorithm export>process contributions for a batching scope</dfn> given
 a [=batching scope=] |batchingScope|, an [=origin=] |reportingOrigin| and a
 [=context type=] |contextType|:
 1. Let |batchEntries| be a new [=list/is empty|empty=] [=list=].
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply spec.bs:396]: I see, in that case, I think it's fine to leave as is to be more clear. Thanks!

File: spec.bs [hunk 9]
```diff
@@ -410,7 +409,8 @@ a [=batching scope=] |batchingScope|, an [=origin=] |reportingOrigin| and a
 1. Let |batchedContributions| be a new [=map/is empty|empty=] [=ordered map=].
 1. [=list/iterate|For each=] |entry| of |batchEntries|:
     1. [=list/Remove=] |entry| from the [=contribution cache=].
-    1. Let |debugDetails| be |entry's| [=contribution cache entry/debug info=].
+    1. Let |debugDetails| be |entry's| [=contribution cache entry/debug
```
[the remark this answers is not in the document]
Comment (alexmturner) [reply spec.bs]: Fixed

File: spec.bs (hunk 6 above)
Comment (alexmturner) [reply spec.bs]: Done

File: spec.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexmturner) [reply spec.bs:396]: Ok, sounds good!

File: spec.bs [hunk 10]
```diff
@@ -297,28 +376,61 @@ To <dfn algorithm export>append an entry to the contribution cache</dfn> given a
...
+
 To <dfn algorithm export>process contributions for a batching scope</dfn> given
 a [=batching scope=] |batchingScope|, an [=origin=] |reportingOrigin| and a
 [=context type=] |contextType|:
 1. Let |batchEntries| be a new [=list/is empty|empty=] [=list=].
 1. [=list/iterate|For each=] |entry| of the [=contribution cache=]:
     1. If |entry|'s [=contribution cache entry/batching scope=] is
-        |batchingScope|, [=list/append=] |entry| to |batchEntries|.
+        |batchingScope|:
+        1. [=Assert=]: |entry|'s [=contribution cache entry/debug details=] is
```
Comment (linnan-github) [inline spec.bs:400]:
```suggestion
        1. [=Assert=]: |entry|'s [=contribution cache entry/debug details=] is not
```

File: spec.bs [hunk 11]
```diff
@@ -203,18 +249,46 @@ General {#general-structures}
...
 A batching scope is a <a spec=HTML>unique internal value</a> that identifies
 which {{PAHistogramContribution}}s should be sent in the same [=aggregatable
-report=].
+report=] unless their [=aggregatable report/debug details=] differ.
 
 Issue: Unique internal value is not an exported definition. See
     <a href="https://github.com/whatwg/infra/issues/583">infra/583</a>.
 
+<h4 dfn-type=dfn>Debug scope</h4>
+A debug scope is a <a spec=HTML>unique internal value</a> that identifies which
```
Comment (alexmturner) [inline spec.bs:258]: Fyi I updated this and the batching scope definition

File: spec.bs (hunk 10 above)
Comment (alexmturner) [reply spec.bs:400]: Ah good catch -- done

**Review by linnan-github — APPROVED** [review]

**Comment by alexmturner** [comment]
Thanks :)

**Pull request merged**